File: library/src/main/java/su/levenetc/android/textsurface/animations/Loop.java
package su.levenetc.android.textsurface.animations;

import su.levenetc.android.textsurface.interfaces.IEndListener;
import su.levenetc.android.textsurface.interfaces.ISurfaceAnimation;

/**
 * Created by ogaclejapan.
 */
public class Loop extends Sequential {

	private final IEndListener restartListener = new IEndListener() {
		@Override public void onAnimationEnd(ISurfaceAnimation animation) {
			if (canceled) {
				if (endListener != null) endListener.onAnimationEnd(Loop.this);
			} else {
				Loop.super.start(restartListener);
			}
		}
	};

	private IEndListener endListener;
	private boolean canceled;

	public Loop(ISurfaceAnimation... animations) {
		super(animations);
	}

	@Override public void start(IEndListener listener) {
		endListener = listener;
		canceled = false;
		super.start(restartListener);
	}

	@Override public void cancel() {
		canceled = true;
		super.cancel();
	}
}

File: library/src/main/java/su/levenetc/android/textsurface/animations/Loop.kt
package su.levenetc.android.textsurface.animations

import su.levenetc.android.textsurface.interfaces.IEndListener
import su.levenetc.android.textsurface.interfaces.ISurfaceAnimation

/**
 * Created by ogaclejapan.
 */
class Loop(vararg animations: ISurfaceAnimation) : Sequential(*animations) {
    private val restartListener: IEndListener = object : IEndListener {
        override fun onAnimationEnd(animation: ISurfaceAnimation) {
            if (canceled) {
                if (endListener != null) endListener!!.onAnimationEnd(this@Loop)
            } else {
                super@Loop.start(this)
            }
        }
    }
    private var endListener: IEndListener? = null
    private var canceled = false
    override fun start(listener: IEndListener) {
        endListener = listener
        canceled = false
        super.start(restartListener)
    }

    override fun cancel() {
        canceled = true
        super.cancel()
    }
}